### PR TITLE
Add bronze snapshots for Newer Assets

### LINKED ISSUE #311: Details about Instance Profile
I am looking for details about  Instance Profile . In what table we can find them ?

### LINKED ISSUE #431: [FEAT] - Add bronze snapshots libraries / clusterPolicies
Create bronze snapshot tables for libraries and clusterPolicies

### LINKED ISSUE #433: [FEAT] -- Add additional snapshots to bronze
Add the following API lists as bronze entities for future consumption in Silver/Gold

- [instance profiles](https://docs.databricks.com/dev-tools/api/latest/instance-profiles.html#list)
- [global init scripts](https://docs.databricks.com/dev-tools/api/latest/global-init-scripts.html#operation/get-scripts)
- [job runs](https://docs.databricks.com/dev-tools/api/latest/jobs.html#operation/JobsRunsList)
  - Use API version 2.1
  - Use APICallV2
  - `expand_tasks=true`
- [tokens](https://docs.databricks.com/dev-tools/api/latest/tokens.html#list)

### PR DESCRIPTION
**Pull request by GeekSheikh**
closes #431 
closes #433 

partially resolves #311

### CODE REVIEW

File: src/main/scala/com/databricks/labs/overwatch/pipeline/Bronze.scala [hunk 1]
```diff
@@ -112,6 +112,18 @@ class Bronze(_workspace: Workspace, _database: Database, _config: Config)
     append(BronzeTargets.poolsSnapshotTarget)
   )
 
+  lazy private[overwatch] val libsSnapshotModule = Module(1007, "Bronze_Libraries_Snapshot", this)
+  lazy private val appendLibsProcess = ETLDefinition(
+    workspace.getClusterLibraries,
+    append(BronzeTargets.libsSnapshotTarget)
+  )
+
+  lazy private[overwatch] val policiesSnapshotModule = Module(1008, "Bronze_Libraries_Snapshot", this)
```
Comment (aman-db) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/Bronze.scala:121]: module name should be different, Bronze_Libraries_Snapshot has been used in the above module(1007)

**Comment by aman-db** [comment]
@GeekSheikh - Ready for review

File: src/main/scala/com/databricks/labs/overwatch/ApiMeta.scala [hunk 2]
```diff
@@ -177,3 +182,28 @@ class ClusterEventsApi extends ApiMeta {
   setApiCallType("POST")
   setStoreInTempLocation(true)
 }
+
+class JobRunsApi extends ApiMeta {
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/ApiMeta.scala:186]: This needs to use API v2.1 or we'll have issues. Please work with @Sriram-databricks to implement. You'll probably need to merge 0612_release branch into yours to pickup the changes you need to make this happen.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineTargets.scala [hunk 3]
```diff
@@ -154,6 +186,24 @@ abstract class PipelineTargets(config: Config) {
...
     )
 
+    lazy private[overwatch] val jobRunsSnapshotTarget: PipelineTable = PipelineTable(
+      name = "job_runs_snapshot_bronze",
+      _keys = Array("job_id", "run_id", "Overwatch_RunID"),
+      config,
+      incrementalColumns = Array("Pipeline_SnapTS"),
+      statsColumns = "created_time, creator_user_name, job_id, run_id, Pipeline_SnapTS, Overwatch_RunID".split(", "),
+      partitionBy = Seq("organization_id"),
+      masterSchema = Some(Schema.jobSnapMinimumSchema)
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineTargets.scala:196]: this is the jobSnapMinSchema -- you should remove this for now as we don't have downstream dependencies so we don't have a master schema for this yet.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineTargets.scala [hunk 4]
```diff
@@ -154,6 +186,24 @@ abstract class PipelineTargets(config: Config) {
       config
     )
 
+    lazy private[overwatch] val jobRunsSnapshotTarget: PipelineTable = PipelineTable(
+      name = "job_runs_snapshot_bronze",
+      _keys = Array("job_id", "run_id", "Overwatch_RunID"),
+      config,
+      incrementalColumns = Array("Pipeline_SnapTS"),
+      statsColumns = "created_time, creator_user_name, job_id, run_id, Pipeline_SnapTS, Overwatch_RunID".split(", "),
```
Comment (GeekSheikh) [inline src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineTargets.scala:194]: do all of these columns actually exist? Do you truly want to move all these to the front and have stats calculated on these columns? If so, no issue, but just wanted to check it's not another copy/paste issue.

**Review by GeekSheikh — COMMENTED** [review]
Please make the changes noted in the review.

File: src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineTargets.scala (hunk 3 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineTargets.scala:196]: removed `Some(Schema.jobSnapMinimumSchema)`

File: src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineTargets.scala (hunk 4 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/pipeline/PipelineTargets.scala:194]: removed  `statsColumns`

File: src/main/scala/com/databricks/labs/overwatch/ApiMeta.scala (hunk 2 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/ApiMeta.scala:212]: added for job runs

**Comment by GeekSheikh** [comment]
@Sriram-databricks -- please approve this PR. I opened it (for some silly reason) so I cannot approve it. LGTM

File: src/main/scala/com/databricks/labs/overwatch/ApiCall.scala [hunk 5]
```diff
@@ -150,6 +150,10 @@ class ApiCall(env: ApiEnv) extends SparkSessionWrapper {
         case "instance-pools/list" => "instance_pools"
         case "instance-profiles/list" => "instance_profiles"
         case "workspace/list" => "objects"
+        case "libraries/all-cluster-statuses" => "statuses"
```
Comment (Sriram-databricks) [inline src/main/scala/com/databricks/labs/overwatch/ApiCall.scala:153]: why APICall??? we should be using ApiCallV2 for all of our enhancement.

File: src/main/scala/com/databricks/labs/overwatch/env/Workspace.scala [hunk 6]
```diff
@@ -130,6 +130,73 @@ class Workspace(config: Config) extends SparkSessionWrapper {
...
+      .execute()
+      .asDF()
+      .withColumn("organization_id", lit(config.organizationId))
+  }
+
+  /**
+   * TODO - Need to find author of the below code
+   *
+   */
+  //  private def clusterState(apiEnv: ApiEnv): String = {
```
Comment (Sriram-databricks) [inline src/main/scala/com/databricks/labs/overwatch/env/Workspace.scala:183]: Do we need this unused commented code?

**Review by Sriram-databricks — CHANGES_REQUESTED** [review]
I see some changes in APICall which we should not use.Please let me know to discuss more.

File: src/main/scala/com/databricks/labs/overwatch/env/Workspace.scala (hunk 6 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/env/Workspace.scala:183]: we need to find the author for this code. Thats why we kept the commented code

File: src/main/scala/com/databricks/labs/overwatch/ApiCall.scala (hunk 5 above)
Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/ApiCall.scala:153]: As per the issue description we want to use APICallV2 for Job runs

Comment (GeekSheikh) [reply src/main/scala/com/databricks/labs/overwatch/ApiCall.scala:153]: @aman-db -- I did not add this. I'm not sure which branch it came in on

Comment (aman-db) [reply src/main/scala/com/databricks/labs/overwatch/ApiCall.scala]: @Sriram-databricks removed the legacy code from APICall.scala

**Review by Sriram-databricks — APPROVED** [review]

**Comment by GeekSheikh** [comment]
@Sriram-databricks @aman-db -- sorry fellas, but this has been rebased to the 0701_release branch. I believe most of the conflicts are fairly trivial -- please take a look and try to get this ready to merge into 0701 resolving all the conflicts. @Sriram-databricks -- I know you are going on vacation -- this is 100% fine to do when you return.

**Pull request closed without merging**